### PR TITLE
Fix ResultProcessor bug, LogisticRegression bug and missing value conversion bug

### LINKED ISSUE #1186: ResultProcessor.cs: code under "#if TLCFULLBUILD" doesn't build
Old code had result.Trainer.Kind, new syntax is result.TrainerKind.

### LINKED ISSUE #1187: Conversion: TryParseKey() throws an exception
The method checks that IsStdMissing() returns false, but it is called without doing that check.
We should remove this check and return false if IsStdMissing().

### LINKED ISSUE #1188: Sweeper: ConfigRunner runs ResultProcessor without loading any assemblies into ComponentCatalog
ExeConfigRunner uses ResultProcessor to summarize the metrics of the different runs, however, ResultProcessor needs to find components in the ComponentCatalog, so they should be loaded before calling it.

### LINKED ISSUE #1205: Logistic Regression doesn't work with training stats
Stats that are being computed at training time aren't passed to the predictor.

### PR DESCRIPTION
**Pull request by yaeldekel**
- LogisticRegression doesn't pass the training stats to the predictor: fixes #1205 .
- ResultProcessor has code that doesn't build: fixes #1186 .
- Missing value conversion: fixes #1187 .
- Sweeper needs to load all components into ComponentCatalog: fixes #1188 .

### CODE REVIEW

File: src/Microsoft.ML.Data/Data/Conversion.cs [hunk 1]
```diff
@@ -1170,7 +1170,7 @@ private bool IsStdMissing(ref ReadOnlySpan<char> span)
         public bool TryParseKey(ref TX src, U8 min, U8 max, out U8 dst)
         {
             var span = src.Span;
-            Contracts.Check(!IsStdMissing(ref span), "Missing text value cannot be converted to unsigned integer type.");
+            Contracts.Check(span.IsEmpty || !IsStdMissing(ref span), "Missing text value cannot be converted to unsigned integer type.");
```
Comment (TomFinley) [inline src/Microsoft.ML.Data/Data/Conversion.cs]: >span.IsEmpty [](start = 28, length = 12)

Heh heh. Whoops!

File: test/data/Mapping.de-de.txt [hunk 2]
```diff
@@ -0,0 +1,7 @@
+Wirtschaft	0
```
Comment (TomFinley) [inline test/data/Mapping.de-de.txt]: This is associated with `lm.sample.txt`, I guess? So I wonder if `lm.labels.txt` might be a better name, just to make clear the two files are in some way associated -- `Mapping.de-de.txt` is not entirely clear.

File: test/Microsoft.ML.TestFramework/TestCommandBase.cs [hunk 3]
```diff
@@ -1154,13 +1154,14 @@ public void CommandTrainMlrWithLabelNames()
             Done();
         }
 
-        [Fact(Skip = "Need CoreTLC specific baseline update")]
+        //[Fact(Skip = "Need CoreTLC specific baseline update")]
+        [Fact]
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.TestFramework/TestCommandBase.cs]: make your mind :)

File: test/BaselineOutput/SingleDebug/Command/CommandTrainingLrWithStats-1-out.txt [hunk 4]
```diff
@@ -0,0 +1 @@
+Saving predictor summary
```
Comment (Ivanidzo4ka) [inline test/BaselineOutput/SingleDebug/Command/CommandTrainingLrWithStats-1-out.txt]: >summary [](start = 17, length = 7)

with latest Eric changes ([https://github.com/dotnet/machinelearning/pull/1193](https://github.com/dotnet/machinelearning/pull/1193)) you can just use Common folder if files for Debug and Release are same.

File: test/Microsoft.ML.TestFramework/DataPipe/TestDataPipe.cs [hunk 5]
```diff
@@ -27,6 +28,159 @@ public sealed partial class TestDataPipe : TestDataPipeBase
...
+                "Black",
+                "Asian-Pac-Islander",
+                "White",
+            });
+
+            string pathData = GetDataPath("adult.train");
+            TestCore(pathData, false,
+                new[] {
+                    "loader=Text{header+ sep=comma col=Label:14 col=Age:0 col=Gender:TX:9 col=Mar:TX:5 col=Race:TX:8 col=Num:2,4,10-12 col=Txt:TX:~}",
+                    "xf=Cat{col=Race2:Key:Race data={" + pathTerms + "} termCol=Whatever}",
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.TestFramework/DataPipe/TestDataPipe.cs]: >Race2:Key:Race [](start = 32, length = 14)

Is this supported?

Comment (yaeldekel) [reply test/Microsoft.ML.TestFramework/DataPipe/TestDataPipe.cs]: Are you referring to the "Race2:Key:Race" syntax? Or to the term data file argument? The first is supported, the second is supported in the latest iteration of this PR :-).

---
In reply to: [224882266](https://github.com/dotnet/machinelearning/pull/1236#discussion_r224882266) [](ancestors = 224882266)

File: test/Microsoft.ML.TestFramework/DataPipe/TestDataPipe.cs [hunk 6]
```diff
@@ -27,6 +29,284 @@ public sealed partial class TestDataPipe : TestDataPipeBase
         private static VBuffer<Double> dataDoubleSparse = new VBuffer<Double>(5, 3, new double[] { -0.0, 0, 1 }, new[] { 0, 3, 4 });
         private static uint[] resultsDoubleSparse = new uint[] { 21, 21, 21, 21, 31 };
 
+        [Fact]
+        public void SavePipeLabelParsers()
+        {
+            string pathData = GetDataPath(@"lm.sample.txt");
+            string mappingPathData = GetDataPath(@"lm.labels.txt");
+
+            // REVIEW shonk: The file doesn't really have a header row. Is it intentional to pretend it does?
```
Comment (sfilipi) [inline test/Microsoft.ML.TestFramework/DataPipe/TestDataPipe.cs]: >shonk [](start = 22, length = 5)

think we don't use aliases here anymore.

File: src/Microsoft.ML.Transforms/CategoricalTransform.cs [hunk 7]
```diff
@@ -212,15 +214,17 @@ internal void SetTerms(string terms)
         /// <param name="outputKind">The type of output expected.</param>
         public CategoricalEstimator(IHostEnvironment env, string name,
             string source = null, CategoricalTransform.OutputKind outputKind = Defaults.OutKind)
-            : this(env, new ColumnInfo(source ?? name, name, outputKind))
+            : this(env, new[] { new ColumnInfo(source ?? name, name, outputKind) })
         {
         }
 
-        public CategoricalEstimator(IHostEnvironment env, params ColumnInfo[] columns)
+        public CategoricalEstimator(IHostEnvironment env, ColumnInfo[] columns,
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Transforms/CategoricalTransform.cs]: >public [](start = 8, length = 6)

out of curiosity can you make this constructor internal? I would prefer to not pollute our public API with these things.
Same for TermEstimator.

**Pull request merged**